### ISSUE TITLE
The new operator no longer reconciles realms, client, users, …

### ISSUE THREAD

**Issue by ctron** [keycloak/keycloak#11928]
### Description

In the past, with the "legacy" operator, it was possible to create instances, realms, clients, and users, which got reconciled so that it was possible to declare a desired state, and let the operator reconcile. Which is a very common pattern with operators.

The new operator only reconciles the instance now, but no longer allows for realms, clients, users to be reconciled. This is a pretty major regression IMHO, as this takes away most of the functionality of an operator.

### Motivation

Operators are there to reconcile the desired state with the current state. Just for installation a Helm chart would be way simpler and more resource efficient.

Also, if you want to deploy a system using Keycloak, having a way to do GitOps style deployments makes a lot of sense. And that would require to use some kind of declarative state configuration. If every user would now be required to create their own operator, that would be a big step back IMHO.

**Comment by andreaTP** [comment]
Hi @ctron ! And thanks for the feedback!

Unfortunately, with the old operator, we faced some pretty major issues with the syncing of resources as it was not properly supported by Keycloak itself.

With the new Operator this functionality is planned to be implemented on top of Store.X that will give a much more solid foundation for it.

cc. @vmuzikar

**Comment by ctron** [comment]
> With the new Operator this functionality is planned to be implemented …

Ok, that sounds we just have to be a bit more patient. Thanks for the update :)

**Comment by philomory** [comment]
Would it be possible to explicitly document this lack of functionality somewhere? I mean, I know there's no documentation stating you _can_ do this with the new operator, but given that you could do it with the old operator, the fact that you can't do it (for now) with the new operator isn't necessarily obvious.

**Comment by andreaTP** [comment]
For context, this has been detailed in [the original design proposal](https://github.com/keycloak/keycloak-community/pull/330/files#diff-1a3bcf2f8669c80d5b09bd9e1f3cd65d516dd36f1857cac43d605c362bca459dR63).

**Comment by vmuzikar** [comment]
@philomory Thanks for trying out the new Operator! It is actually [documented](https://www.keycloak.org/migration/migrating-to-quarkus#_migrating_using_the_operator) (see the last paragraph):
> The new operator doesn’t support Client, User and Realm CRDs directly. Instead, it provides one CRD to perform a [Realm import](https://www.keycloak.org/operator/realm-import.html). Using this new CR you can import Users, Clients and more through the wrapping Realm.

**Comment by mjschmidt** [comment]
https://www.keycloak.org/docs-api/18.0.0/rest-api/index.html#_realmrepresentation
This link does not work btw

**Comment by andrewm-aero** [comment]
I've just today discovered the existence of this new operator, and I have to echo ctron's sentiment. What is the point of the operator if managing Keycloak objects via the K8s api is not a goal? Unless I'm completely misunderstanding the realm-import page incorrectly, the CR is inspected once, and then never reconciled again. In that case, why not just make a standard Keycloak API call? The whole purpose of an operator is to make a declarative API for an otherwise imperative API. What justification is there to use this over, say, a helm chart? Looking at https://github.com/keycloak/keycloak/blob/main/operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java, there doesn't appear to be any behavior that can't be replicated by one. Reading the original proposal linked, the idea of mounting a CR as YAML stored in git for read-only configuration actually sounds very close to how my team has been operating for the last few years, using ArgoCD to store static configuration (clients, user roles, etc) in git using the (now) legacy operator CRs, while using Postgres for nothing more than password and client secret storage, and treating the admin console as read-only. Is this still a use case that is intended to be supported, or is the one-time realm import going to remain the end-all and be-all of Keycloak API interaction through CRs? If nothing else, I'd like to draw attention to the smaller-than-you-expect maximum size of k8s resources, so particularly large Realms would outright not work here.

**Comment by vmuzikar** [comment]
@andrewm-aero Thank you for your interest in the new Operator and raising the concerns. It is a valid point that the Operator as is now no "smarter" than a Helm chart. However, the Operator is still in development and its current state is far from something we could consider final (seamless Keycloak upgrades, auto-scaling etc.). The plan also includes adding more CRs (probably Realm and Client at minimum) and leverage the new static store to apply them in Keycloak in fully declarative way. But since the static store is not yet ready, we wanted to offer at least a stop-gap solution – Realm Import CR.

Based on our previous experience with the legacy Operator, we decided not to use the Keycloak Admin REST API for managing the resources in the new Operator. It turned out to be fragile and not really reliable due to complexity and dependencies between endpoints, resulting in high maintenance costs.

Thank you for your understanding and patience. Stay tuned for news around the new Operator and Keycloak.X. :)

**Comment by andrewm-aero** [comment]
Thanks for the quick response, that's certainly encouraging. I don't want to tell you how to do your job, but I think this might be an issue of messaging, and you might want to coordinate with the people managing the legacy operator, since their page says its been deprecated that people should be migrating to the new operator _now_, which, if its not ready yet, leads to confusion like mine. Maybe adjusting the wording on that, and linking directly to a page describing "What can new operator do today" as well as "What new operator is planned to do" might save you guys some headaches dealing with similarly confused users. As far as I can tell, that original design document is on a detached commit, so its not super easy to find.

I look forward to seeing what this new operator becomes.

**Comment by ctron** [comment]
Following up on:

> Adding more CRDs is definitely on our radar

And:

> However, the Operator is still in development and its current state is far from something we could consider final 

I think in order to clearly communicate the state of this, either this issue should be kept open, and regularly updated. Or, even better, there should be a very prominent (pinned) tracking issue, with a todo style list, referencing open issues.

The reason for that is, as already mentioned and agreed by others in this issue, that this is a pretty existential issue of the operator. So it is helpful for people to understand when is the right time to really invest into adopting the new operator.

**Comment by KarstenSiemer** [comment]
Hello my friends,
first of all, I'd like thank everyone involved with all the work they do on this open-source project. I am truly thankful for the ecosystem that was created and the passion that people invest.

Yet, I hope to find some comrades-in-arms to my troubles here.

I have just implemented a keycloak-operator installation and I am confused with all the different types of operators and config-cli tools.

I am satisfied with the keycloak cluster the quarkus operator created for me.

I now desire:
* auto create/update/delete of realms/users/clients using crds
* Beste case would be split management where I can create users outside of the realm crd with full reconciliation

Using the quarkus operator, I can create a realm with users and clients, but cannot manage it.

By googling and looking through other peoples issues, I found that the old operator was archived and forked to a new old operator, which is the thing as the old operator minus keycloak instance management.

On the new old operators readme page, it is said that starting keycloak 21, the quarkus operator will gain crds for users and clients and realms will be manageable though their whole lifecycle.
Keycloak 21 is realeased now, yet I fail to find any updates or status on that.
But on the new old operators page it is also said to create the realm using the quarkus operator and reference the quakus keycloak instance as an external keycloak to gain full realm representation. As far as I understand, the realm should at best not include any user/clients except for ones I am sure are full static - as this Realm created using the RealmImport crd will not be managed after the initial creation.
Next I can manage user/clients using the new old operators crds.
The main problem I have with this is https://github.com/keycloak/keycloak-operator/issues/499 (problem persisted in new old operator), I have to add the users secret cleartext in the crd which is completely unacceptable. We are using a GitOps approach and no controller is in place that could do an in-time replacement as ArgoCD directly pulls the resources stored in git. So that is really out of the window for me.

So the only thing left for me, I guess, is using https://github.com/adorsys/keycloak-config-cli , as I can deploy it to kubernetes as well as a pod and use kustomize to bundle realms and users. The users password can be stored in an encrypted way, which does solve the problem ( to a great extend) I have with the new old operators user crd, which does only support clear text unsupported passwords. If the new old operator would support reading the users secret from a preexisting secret, it would be fine, as I could present that using sealed secrets or something else.


So now my big question, are my assumptions correct and I should proceed the way I planned or am I missing something?
Is there an update when extended lifecycle management of resources will be supported by quarkus operator, or some Issue/page I can look into to either support or watch the issue in general?
